viper/parser/parser.py

```diff
@@ -277,13 +277,31 @@ def __init__(self, vars=None, globals=None, sigs=None, forvars=None, return_type
         self.in_for_loop = set()
         # Count returns in function
         self.function_return_count = 0
+        # Current blockscope
+        self.blockscopes = []
 
     def set_in_for_loop(self, name_of_list):
         self.in_for_loop.add(name_of_list)
 
     def remove_in_for_loop(self, name_of_list):
         self.in_for_loop.remove(name_of_list)
 
+    def start_blockscope(self, blockscope_id):
+        if blockscope_id not in self.blockscopes:
+            self.blockscopes.append(blockscope_id)
+
+    def end_blockscope(self, blockscope_id):
+        if blockscope_id not in self.blockscopes:
+            return
```

Comment:
Shouldn't this raise an error?

Reply:
I could probably remove that check completely?

Reply:
Maybe, I mean this is an internal thing you added to keep track of blockscopes, which is something Python already does with the AST module to prevent uses from changing it arbitrarily.

Wait, does AST provide any information for this?

Reply:
Not that I am aware no, since python doesn't use block scoping like we want to. I'll remove it - just a bit of 'defensive coding' ;)

Reply:
I don't think AST will help here as it only parses with python scoping.  I think this check can be removed given that we're only calling it internally and inputting `blockscope_id`'s that will be in `self.blockscopes`

Reply:
Python does scoping, at least behind the scenes. All languages do some aspect of scoping. Python specifically tracks indenting to figure out block scoping. This might be handled in the lexer and not made present in the AST, but it's definitely in the Python front-end parser somewhere.

Reply:
You need to pop scopes from the list to prevent this from compiling:
```python
def foo(b: bool):
    if b:
        a: num = 1
    a += 2
```

Reply:
@jacqueswww I don't think the above example will compile.  Does it currently?

Reply:
Did for me, will be reworking this PR today ;)